text/0000-target-feature-runtime.md

```diff
@@ -0,0 +1,457 @@
+time. We could implement this as proposed, but only stabilize using the
+target-feature detection macros via `libcore`. This would mean that initially,
+`#![no_std]` binaries won't be able to implement their own run-times, but that
+would unlock using the macros on all `#![no_std]` libraries, and these macros
+would do something meaningful if `libstd` is linked into the final binary.
+
+## libcore pulls target-features approach
+
+We could provide a "cache" in libcore, and an API for users or only for the
+standard library, to initialize this cache externally, e.g., during the standard
+library initialization routine.
+
+This runs into problems with `cdylib`s, where these routines might not be called
+automatically, potentially requiring C code to have to manually call into
+`libstd` initialization routines. It also runs into problems with often imposing
+a cost on users, e.g., due to a cache in libcore, even though users might never
+use it. This would be limiting, if e.g. having a cache in read-write memory
+prevents libcore from being compiled to a read-only binary. We would need to
+feature gate this functionality to avoid these issues.
+
+It isn't cross-domain either, e.g., an OS kernel would need to disable this
+functionality, and wouldn't be able to provide their own. So while they could
+use libraries that would do run-time feature detection, no meaningful detection
+would be performed.
+
+# Prior art
+[prior-art]: #prior-art
+
+This feature is very similar to `#[global_allocator]` and `#[panic_handler]`.
+Since a default implementation is provided if the user does not provide one,
+this is a backward compatible change.
+
+This feature does not exist in any programming languages I know. Clang and GCC
+do have a feature-detection run-time, which is not configurable, nor does it
+work for all users.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* We could implement this RFC, without making any APIs public, by just moving
+  the feature-detection macros to `libcore`. That would allow `#![no_std]`
+  libraries to use them, and they will do something meaninful if `libstd` is
+  linked. `#![no_std]` binaries won't be able to provide their own run-time, but
+  the APIs for this (the trait, enum, and `#[target_feature_detection_runtime]`
+  attribute) could be stabilized at a later time.
+
+* Should the API use a `TargetFeature` `enum` or be stringly-typed like the
+  macros and use string literals?
+  
+* Since the `TargetFeature` `enum` is architecture-specific, should it live in
+  `core::arch::{target_arch}::TargetFeature` ?
+
+* How does it fit with the Roadmap? Does it fit with the Roadmap at all? Would
+  it fit with any future Roadmap?
+
+* Should the `libstd` run-time be overridable? For example, by only providing it
+  if no other crate in the dependency graph provides a runtime ? This would be a
+  forward-compatible extension, but no use case considered requires it.
+
```

Comment:
* Does `core::detect::is_target_feature_detected` cache the result of `core::detect::TargetFeatureRuntime::is_feature_detected` or is the later responsible for caching itself?